Authorization and routing(HW2(FE))

zFarrukh opened the pull request.

zFarrukh committed:
add cors to be able to connect frontend with backend

zFarrukh committed:
add authorization routing 404 page auth guard

zFarrukh committed:
update user controller

zFarrukh committed:
add guard to category route

zFarrukh committed:
implement sharedModule

zFarrukh committed:
add logo icon

vokanred commented on client/src/app/app-routing.module.ts:
```diff
@@ -1,10 +1,16 @@
+import { NotFoundComponent } from './not-found/not-found.component';
 
-const routes: Routes = [];
+const routes: Routes = [
```
Don't you have any routes?

zFarrukh replied:
Yeah I modified this file and use lazy loading

vokanred commented on client/src/app/auth/auth-form/auth-form.component.ts:
```diff
@@ -0,0 +1,41 @@
+
+  public onFormSubmit(): void {
+    const { email, password } = this.loginForm.value;
+    this.authService.login(email, password).subscribe({
```
Unsubscribe on destroy

vokanred commented on client/src/app/auth/auth-form/auth-form.component.ts:
```diff
@@ -0,0 +1,41 @@
+    const { email, password } = this.loginForm.value;
+    this.authService.login(email, password).subscribe({
+      next: (userData) => {
+        console.log(userData);
```
Remove

vokanred commented on client/src/app/auth/auth-form/auth-form.component.ts:
```diff
@@ -0,0 +1,41 @@
+    this.authService.login(email, password).subscribe({
+      next: (userData) => {
+        console.log(userData);
+        this.loginErrorMessage = '';
```
It's better to clear error message before sending request

vokanred commented on client/src/app/auth/auth.module.ts:
```diff
@@ -0,0 +1,25 @@
+  imports: [
+    CommonModule,
+    ReactiveFormsModule,
+    RouterModule.forChild(routes),
```
Usually forChild() is used for lazy loaded modules

zFarrukh replied:
Updated

vokanred commented on client/src/app/auth/auth.module.ts:
```diff
@@ -0,0 +1,25 @@
+    RouterModule.forChild(routes),
+    SharedModule,
+  ],
+  exports: [AuthFormComponent],
```
In the current state of your code you don't need this export. But I'd recommend to keep the export and move /login route to app routing.

vokanred commented on client/src/app/auth/services/auth.service.ts:
```diff
@@ -0,0 +1,53 @@
+
+  login(email: string, password: string): Observable<User> {
+    return this.http
+      .post('http://localhost:3000/user/login', {
```
Move API url to environment

vokanred commented on client/src/app/auth/services/auth.service.ts:
```diff
@@ -0,0 +1,53 @@
+        password,
+      })
+      .pipe(
+        tap((res: User | any) => {
```
Do not use any

zFarrukh replied:
I tried but it is giving me error

zFarrukh replied:
I fixed it

vokanred commented on client/src/app/auth/services/auth.service.ts:
```diff
@@ -0,0 +1,53 @@
+
+  private setSession(data: User): void {
+    localStorage.setItem('token', data.token);
+    localStorage.setItem(
```
If you want to store expiresIn pls get it from API response because expiration time is set on the server using env variable

vokanred commented on client/src/app/auth/services/auth.service.ts:
```diff
@@ -0,0 +1,53 @@
+    if (!token) {
+      return false;
+    }
+    const expiresIn = localStorage.getItem('expiresIn');
```
You don't need to check if token is expired. If this is happened API will automatically return 401

vokanred commented on client/src/app/auth/auth.interceptor.ts:
```diff
@@ -0,0 +1,28 @@
+  constructor(private authService: AuthService) {}
+
+  intercept(
+    req: HttpRequest<any>,
```
Use unknown instead of any

vokanred commented on client/src/app/auth/auth.module.ts:
```diff
@@ -0,0 +1,25 @@
+
+const routes: Routes = [
+  {
+    path: 'login',
```
I think you should protect this page for authorized users

vokanred reviewed (changes requested):
I've not understood the routing in the app. Pls review my comments

zFarrukh committed:
add lazy loading routing

zFarrukh committed:
unsubscribe ngOnDestroy on auth-form component

zFarrukh committed:
fix bugs modify files according to review

zFarrukh committed:
fixed issue

zFarrukh commented:
Hi, I modified files according to your review, waiting for your approve

vokanred commented on client/src/app/app-routing.module.ts:
```diff
@@ -1,10 +1,38 @@
+const routes: Routes = [
+  {
+    path: 'category',
+    pathMatch: 'full',
```
You don't need pathMatch here

vokanred commented on client/src/app/app-routing.module.ts:
```diff
@@ -1,10 +1,38 @@
+  },
+  {
+    path: 'login',
+    pathMatch: 'full',
```
And here

vokanred commented on client/src/app/app-routing.module.ts:
```diff
@@ -1,10 +1,38 @@
+    component: AuthFormComponent,
+  },
+  {
+    path: '',
```
But I would add pathMatch here

zFarrukh replied:
I have statistics route inside mainModule

vokanred reviewed (changes requested):
good job. pls review comments
and what about expiration date? don't you agree?

zFarrukh committed:
remove pathMatch from routes

zFarrukh committed:
add angular-jwt and use it to check tokenIsExpired and add authorization header to requests by jwtModule remove auth interceptor

zFarrukh commented:
Hi, I modified it and used angular-jwt to check if the token is expired or not and removed auth interceptor and instead used jwtModule from the angular-jwt library

vokanred reviewed (approved):
Done
